arm9/source/game/cert.c

```diff
@@ -1,21 +1,319 @@
+    if (!cache) return 1;
+    if (BuildDisaDiffDpfsLvl2Cache(path, &info, cache, info.size_dpfs_lvl2) != 0) {
+        free(cache);
+        return 1;
+    }
+
+    Certsdb_partition_header header;
+
+    if (ReadDisaDiffIvfcLvl4(path, &info, 0, sizeof(Certsdb_partition_header), &header) != sizeof(Certsdb_partition_header)) {
+        free(cache);
+        return 1;
+    }
+
+    if (getbe32(header.magic) != 0x43455254 /* 'CERT' */ ||
+      getbe32(header.unk) != 0 ||
+      getle32(header.used_size) & 0xFF) {
+        free(cache);
+        return 1;
+    }
+
+    u32 offset = sizeof(Certsdb_partition_header);
+    u32 max_offset = getle32(header.used_size) + sizeof(Certsdb_partition_header);
+
+    u32 ret = 1;
+
+    // certs.db has no filesystem.. its pretty plain, certificates after another
+    // but also, certificates are not equally sized
+    // so most cases of bad data, leads to giving up
+    while (offset < max_offset) {
+        char full_issuer[0x41];
+        u8 sig_type_data[4];
+        u8 keytype_data[4];
+
+        if (offset + 4 > max_offset) break;
+
+        if (ReadDisaDiffIvfcLvl4(path, &info, offset, 4, sig_type_data) != 4)
+            break;
+
+        u32 sig_type = getbe32(sig_type_data);
+
+        if (sig_type == 0x10002 || sig_type == 0x10005) break; // ECC signs not allowed on db
+
+        u32 sig_size = _Certificate_GetSignatureChunkSizeFromType(sig_type);
+        if (sig_size == 0) break;
+
+        u32 keytype_off = offset + sig_size + offsetof(Certificate_body, keytype);
+        if (keytype_off + 4 > max_offset) break;
+
+        if (ReadDisaDiffIvfcLvl4(path, &info, keytype_off, 4, keytype_data) != 4)
+            break;
+
+        u32 keytype = getbe32(keytype_data);
+
+        if (keytype == 2) break; // ECC keys not allowed on db
+
+        u32 data_size = _Certificate_GetDataChunkSizeFromType(keytype);
+        if (data_size == 0) break;
+
+        u32 full_size = sig_size + data_size;
+        if (offset + full_size > max_offset) break;
+
+        cert_local.sig = (Certificate_signature*)malloc(sig_size);
+        cert_local.data = (Certificate_body*)malloc(data_size);
+        if (!cert_local.sig || !cert_local.data)
+            break;
+
+        if (ReadDisaDiffIvfcLvl4(path, &info, offset, sig_size, cert_local.sig) != sig_size)
+            break;
+
+        if (ReadDisaDiffIvfcLvl4(path, &info, offset + sig_size, data_size, cert_local.data) != data_size)
+            break;
+
+        if (!Certificate_IsValid(&cert_local))
+            break;
+
+        if (snprintf(full_issuer, 0x41, "%s-%s", cert_local.data->issuer, cert_local.data->name) > 0x40)
+            break;
+
+        if (!strcmp(full_issuer, issuer)) {
+            ret = 0;
+            break;
+        }
+
+        free(cert_local.sig);
+        free(cert_local.data);
+        cert_local.sig = NULL;
+        cert_local.data = NULL;
+
+        offset += full_size;
+    }
+
+    if (ret) {
+        free(cert_local.sig);
```

Comment:
Same here, it's cleaner to use the function you wrote.

Reply:
Also fixed with commit d346524